feat: Add async functionality

Were-Woof opened the pull request:
- Add ``AsyncClass`` to allow asynchronous use of the api
- Add unittests to test ``AsyncClass``
- Add ``AsyncClass`` to the sphinx documentation

closes #3

Were-Woof committed:
feat(asyncClient): Add untested AsyncClient

Were-Woof committed:
fix: Rewrite methods to prevent closed ClientSession from being used

Were-Woof committed:
fix(docstrings): Update docstrings to display current parameters and returns

Were-Woof committed:
docs: Add AsyncClient to the sphinx docs

Were-Woof committed:
feat: Make AsyncClient initialize correctly regardless of if its in an event loop

Were-Woof committed:
fix: Update api endpoint of outputs in AsyncClient

Were-Woof committed:
tests: Add unittests for AsyncClient

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_validate_key_valid_key_raises_no_exception(self, mock_get):
+        mock_get.return_value = {}, 200
```
```suggestion
        mock_get.return_value = {"success": True, "user_id": 0}, 200
```

Incorrect return JSON, the API returns:

```json
{
    "success": true,
    "user_id": 1234567890
}
```

Please fix this otherwise the test should fail.

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+        self.addCleanup(patcher.stop)
+        self.mock_request = patcher.start()
+        self.mock_request.return_value = {}, 200
+        self.client = AsyncClient('valid_key')
```
```suggestion
        self.client = AsyncClient('valid_key', logger=LoggingObject(level=logging.CRITICAL))
```

Logs from the application itself shouldn't show unless it's a error.

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_get_player_returns_player_object(self, mock_get):
+        mock_get.return_value = {'user_id': 1,
```
You shouldn't only have two fields as the response from the request, as it is always going to return data stated on our API documentation. You can obtain mock data/information about the data that is going to be returned [here](https://api.willofsteel.me/#tag/Player-Data/operation/get_player_data_player_get).

Even though this would return a Player object, that is not what we are testing.

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_get_alliance_not_in_alliance_raises_NotInAlliance_Exception(self, mock_get):
+        mock_get.return_value = {'detail': 'Not in an alliance'}, 400
```
```suggestion
        mock_get.return_value = {"success": False, "detail": "alliance not found"}, 400
```

API does not return "Not in an alliance" as the detail, the function just checks for the status code, but the error messages should be correct.

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_update_alliance_name_valid_name_returns_True(self, mock_get):
+        mock_get.return_value = {}, 200
```
```suggestion
        mock_get.return_value = {"updated": "name", "old": "old name", "new": "new name"}, 200
```

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_get_all_offers_invalid_offer_type_raises_KeyError_Exception(self, mock_get):
+        mock_get.return_value = {'detail': 'Invalid offer type'}, 200
```
```suggestion
        mock_get.return_value = {'success': False, 'detail': 'invalid order type'}, 400
```

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,142 @@
+
+    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
+    async def test_get_offer_invalid_offer_type_raises_InvalidInput_Exception(self, mock_get):
+        mock_get.return_value = {'detail': 'Invalid offer type'}, 200
```
```suggestion
        mock_get.return_value = {'success': False, 'detail': 'invalid order type'}, 400
```

ItsNeil17 reviewed (changes requested):
Hey @Were-Woof,

I have requested some minor changes to be made in the test file of the async client. For the changes that I think might require an explaination, I have provided one below the change.


I have not looked at your async client file yet, as all of the tests work and it's getting quite late. I will have a look through it tomorrow to see if I want any more changes, for now the test file is the only file that needs to be updated.

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,319 @@
+        if method not in ["GET", "POST"]:
+            return KeyError("Invalid Method")
+
+        async with aiohttp.ClientSession() as session:
```
Huge blunder.

We do not want to create another session on every request, rather create a session on initialization and use it till the program is killed. 

Please try to accomplish this functionality asap

ItsNeil17 reviewed (changes requested):
I just had a very quick look at the async client file, and saw this huge issue that needs to be taken care of, which is why I'm making another review for this.

Were-Woof committed:
fix(efficiency): Make AsyncClient reuse only one ClientSession throughout its lifetime

Were-Woof committed:
fix: Add ability to cleanup AsyncClient._session when its finished use

Were-Woof committed:
fix(tests): Fix unittests for AsyncClient to use correct mock api data

Were-Woof commented:
Hey @ItsNeil17,

I believe I've addressed all of the issues above. To prevent issues caused by defining a ``ClientSession`` outside of an asynchronous environment, along with issues of leaving it potentially unclosed, I've created a method to handle all asynchronous setup that gets automatically run, as well as adding a method to cleanup any open resources.


I've adjusted the tests as according to your advice, however I did have to change the data return for ``get_player()`` slightly. When using the test api key, it returns:
```
{'success': True, 'user_id': 707274479751135243, 'registered_at': '2023-08-05 01:30:17', 'gold': 0, 'ruby': 0, 'silver': 0, 'food_stored': 133000, 'units': {'infantry': 999999999, 'cavalry': 999999999, 'artillery': 999999999, 'assassins': 999999999, 'bowmen': 999999999, 'big_bowmen': 999999999, 'heavy_men': 999999999, 'kings_guards': 999999999}, 'npc_level': 1, 'last_npc_win': None, 'votes': 0, 'queue_slots': 1000, 'observer': True, 'peace': 1, 'letter_bird': True, 'prestige': 5, 'alliance': {'owner': 702385226407608341, 'created_at': '2023-08-06T11:15:00', 'name': 'Gods allies', 'user_limit': 50, 'bank': 266498586376}}
```

In the documentation and your example above, it returns ``user_id`` as just ``id``, which is not the actual return from the api, it might be outdated?
![image](https://github.com/user-attachments/assets/027e0b2f-769d-4cd8-a203-c0bd4f7e40d2)

Were-Woof committed:
Fix(AsyncClient):

- Make tests for _verify_key use correct data
- Make _verify_key more strict in its checks

ItsNeil17 commented on tests/test_async_client.py:
```diff
@@ -0,0 +1,180 @@
+        asyncio.run(self.client.close())
+        super().tearDown()
+
+    @ patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
```
```suggestion
    @patch('willofsteel.async_client.AsyncClient.request', new_callable=AsyncMock)
```

Please don't add a space between @ and the patch decorator

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+        self._session = aiohttp.ClientSession()
+        await self._verify_key()
+
+    async def close(self):
```
This should be done automatically and not need to be ran by the user. You can use the `__aexit__` function to implement this. More info at https://docs.python.org/3/library/atexit.html

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+            logging.info("Key verification successful.")
+        else:
+            raise ServerError
+
```
Add helper methods such as `get_user_id` (Returns the user ID returned by the API on the verification call), `is_verified` (can be used just incase, returns bool).

These would just be QoL methods for any developers if they need them.

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+            print(data["detail"])
+            print("This error was not automatically detected, please report this to the maintainers (or fix it yourself)!")
+
+    async def get_alliance(self) -> dict:
```
```suggestion
    async def get_alliance(self) -> Alliance:
```

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+        """
+        Retrieve alliance information.
+
+        Returns
+        -------
+        Optional[:class:`~willofsteel.types.Alliance`]
+
```
```suggestion
```

Not optional as it will return an alliance or raise NotInAlliance

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+        new_name: :class:`str`
+            The new name of the alliance.
+
+        Returns
+        -------
+        :class:`bool`
+
```
```suggestion
```

Sphinx automatically takes the return value if it's annotated, but you can keep this if necessary, just a suggestion.

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,333 @@
+        """
+        if len(new_name) > 32:  # this is not an official limit. just a wrapper limit for now
+            raise LimitExceeded(32, "name")
+        headers = self.headers
```
```suggestion
        headers = self.headers.copy()
```

Create a copy of headers, not edit them.

ItsNeil17 reviewed (changes requested):
I have asked for some changes again. Also, I'm really sorry about the mis-match between the documentation and the API itself, I have fixed it to return `user_id` and not `id`.

Were-Woof committed:
feat: Make AsyncClient automatically resolve open resources

Were-Woof committed:
fix: Remove space between @ and patch within tests, remove obsolete tearDown method

Were-Woof committed:
docs(AsyncClient): Update docstrings and type annotations for AsyncClient

Were-Woof committed:
feat: Add quality of life helper functions to AsyncClient: get_user_id(), is_verified()

Were-Woof commented:
Hi,

I've used ``atexit`` to automatically close the ``ClientSession`` and added ``__aexit__`` and ``__aenter__`` so all open resources should now be closed automatically without the user having to do anything.

The docstrings return values inside of the docstrings are not required, I just have them there because I find it easier to read with intellisense, I've left them for now but I have fixed the incorrect docstrings and type annotations.

``is_verified()`` and ``get_user_id()`` have both been added as helper methods, however I was wondering potentially considering them as properties instead? This way we can prevent the user from changing them/reading them if the key fails to verify in the event that they use ``try/except`` to get around the exception it throws.

Also sorry about the ``@ patch`` error, I had my formatter misconfigured.

ItsNeil17 commented:
Agreed, we can make them properties of the client rather than functions. You can implement this if you want @Were-Woof, I'll do my review now.

ItsNeil17 commented on willofsteel/async_client.py:
```diff
@@ -0,0 +1,384 @@
+        """
+        if new_limit > 50:
+            return KeyError  # return an error which says it cant be more than 50
+        headers = self.headers
```
```suggestion
        headers = self.headers.copy()
```

ItsNeil17 reviewed (changes requested):
You must have accidentally left this part unedited, please change this, and I believe we'd be able to merge this after that.

Were-Woof committed:
fix: Make update_alliance_name copy AsyncClient.headers instead of pointing to it

Were-Woof committed:
feat: Convert user_id and is_verified to properties of AsyncClient

Were-Woof commented:
As it turns out I just forgot to commit this change, sorry about that: [323e79a](https://github.com/WillofSteel-Devs/api-wrapper/pull/4/commits/323e79aaaa5e1ee1acf03bcbcc978635bee65de0), it should be ``headers = self.headers.copy()`` now.

I've also added ``user_id`` and ``is_verified`` as properties with some protections to stop users from modifying them, unfortunately python doesn't have strict protections but it should be enough to stop accidental changes: [5bd473e](https://github.com/WillofSteel-Devs/api-wrapper/pull/4/commits/5bd473e97106219d49d01021495d60282e45cbee)

ItsNeil17 reviewed (approved):
LGTM
